### ISSUE TITLE
`torch.linalg.norm` returns `inf` for norms well inside the dtype range, and its gradient is silently `0` (float32/float64)

### ISSUE THREAD

**Issue by shadow-orange41** [pytorch/pytorch#193006]
### 🐛 Describe the bug

`torch.linalg.norm` computes `sqrt(sum(x**2))` directly, so whether the call overflows is
decided by the intermediate **sum of squares**, not by `‖x‖`. A vector whose norm is `1e20`
in float32 — five orders of magnitude below the float32 max of `3.4e38` — returns `inf`,
because `sum(x**2)` is `1e40`.

The gradient is the more damaging half. `backward()` evaluates `x / ‖x‖` as `x / inf` and
produces **`0.0` for every entry**, with no `inf` or `nan` anywhere in the gradient to
signal that anything went wrong. `torch.nn.functional.normalize`, which is built on the
same reduction, likewise returns an all-zero vector rather than a unit vector.

`scipy.linalg.norm` returns the correct value for the identical input, so this is not an
inherent limit of the dtype.

## Reproducer

```python
import torch, numpy as np, scipy.linalg

# true ||v|| = 1e20, well inside float32 range (max 3.4e38).
# Only the intermediate sum(v^2) = 1e40 overflows.
v = np.array([1e20, 1.0, 1.0], dtype=np.float32)
x = torch.tensor(v, requires_grad=True)
y = torch.linalg.norm(x); y.backward()

print("true ||v||        : 1e+20      grad [1.0, 1e-20, 1e-20]")
print("torch.linalg.norm :", float(y))
print("torch grad        :", x.grad.numpy())
print("scipy.linalg.norm :", scipy.linalg.norm(v))
```

### Actual output

```
true ||v||        : 1e+20      grad [1.0, 1e-20, 1e-20]
torch.linalg.norm : inf
torch grad        : [0. 0. 0.]
scipy.linalg.norm : 1.0000000200408773e+20
```

## Affected entry points

Nine float32 entries of `1e19`, as a vector and as a 3×3 (true norm `3e19` in both cases):

```python
v = torch.full((9,),  1e19, dtype=torch.float32)
m = torch.full((3, 3), 1e19, dtype=torch.float32)
```

```
torch.linalg.vector_norm(v)        -> inf
torch.linalg.matrix_norm(m)        -> inf
torch.nn.functional.normalize(v)   -> 0.0   (expected 0.3333)
```

`torch.linalg.norm` and `torch.norm` dispatch to the two primitives above, so they return
`inf` for every `ord=None` / `'fro'` / `ord=2` spelling.

## Note on float16

float16 is **not** affected: `linalg.vector_norm` accumulates low-precision inputs in
float32 (cf. #141953), which is enough headroom that no float16 input with a representable
norm can overflow the accumulator. float16 is correct for exactly the reason float32 and
float64 are not — there is no wider accumulation type in play for them. Passing float32
input through a float64 accumulation manually (`torch.linalg.norm(x.double()).float()`)
returns `3.0e19` correctly, confirming the reduction is where the range is lost.

### Versions

Collecting environment information...
PyTorch version: 2.14.0.dev20260809+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-134-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      41%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.77
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.14.0.dev20260809+cpu
[pip3] torchvision==0.29.0.dev20260810+cpu
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] torch                     2.14.0.dev20260809+cpu          pypi_0    pypi
[conda] torchvision               0.29.0.dev20260810+cpu          pypi_0    pypi

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

**Comment by dhruvildave235** [comment]
I'd like to work on this.

Root cause matches what you've described: `sqrt(sum(x**2))` overflows in the intermediate `sum(x**2)` step whenever any element's square exceeds the dtype's representable range, even though the final norm itself is well within range. This is a well-known failure mode for naive norm computation, and the standard fix (used in LAPACK's `dnrm2`, for example) is to rescale by the max absolute value before squaring:

```python
s = x.abs().max()
norm = s * (x / s).square().sum().sqrt()
```

This avoids the overflow in the reduction step while staying numerically exact for well-scaled inputs, and it should fix the gradient issue too, since `backward()` would then be differentiating through a reduction that never actually produces `inf`.

A few things I want to confirm before opening a PR:
1. Should this scaling be applied unconditionally, or only as a fallback when the naive path would overflow (to avoid any extra overhead on the common case)?
2. Since `vector_norm` and `matrix_norm` are the two primitives everything else dispatches to, should the fix live in one shared helper, or does `matrix_norm` need separate handling for e.g. the Frobenius vs. spectral norm cases?
3. Should I also update `torch.nn.functional.normalize` directly, or does fixing the underlying `vector_norm` primitive it's built on cover that automatically?

Happy to put up a PR once I have a bit more direction on the above.

**Comment by dhruvildave235** [comment]
Following up with what I found tracing the code:

Re #2 (one place or two): mostly one place. linalg_matrix_norm dispatches
differently depending on ord:
- ord=2 (spectral norm) goes through SVD (linalg_svdvals) — a different
  code path entirely, not affected by this bug.
- ord="fro" (Frobenius) calls linalg_vector_norm(A, 2, ...) directly, so
  fixing vector_norm fixes Frobenius matrix norm automatically.
- ord=1/-1/inf/-inf use vector_norm with ord=1 (sum of absolute values,
  no squaring, no overflow risk).
So the fix really only needs to happen in vector_norm's kernel.

Re #3 (normalize): confirmed, torch.nn.functional.normalize is built
directly on vector_norm, so fixing the primitive covers it automatically.

Re #1 (unconditional vs fallback): I traced this down to the actual CPU
kernel — aten/src/ATen/native/cpu/ReduceOpsKernel.cpp has a vectorized
SIMD fast path (norm_two_reduce_step) that does the sum-of-squares, and
there's likely a second implementation in SharedReduceOps.h (NormTwoOps)
for the general/fallback case, plus a probably-identical bug in the CUDA
kernel. I don't have a local build or GPU to benchmark the rescaling
overhead against these SIMD paths, so I don't think I'm the right person
to answer the performance question or safely modify the vectorized code
myself. Happy to hand off the fix from here if that's useful, or pair
with a maintainer who can validate the perf side.

**Comment by shadow-orange41** [comment]
Thanks for looking into this, @dhruvildave235! Please feel free to go ahead and open a PR.

**Comment by shadow-orange41** [comment]
This shows the well-known numerical issue with `norm`: `sum(x^2)` grows quickly and can overflow, and lower-precision types make this worse simply because their representable range is smaller (see Higham, [Accuracy and Stability of Numerical Algorithms](https://epubs.siam.org/doi/10.1137/1.9780898718027), §27.8, chapter screenshot attached).

<img width="1593" height="899" alt="Image" src="https://github.com/user-attachments/assets/7ebd94be-b54e-447f-81f0-f7f764347abd" />

I was checking how PyTorch handles this. norm accumulates using the `acc_t` type determined by `at::opmath_type`, then casts the result back to the original dtype ([ReduceOpsKernel.cpp#L189-L190](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L189-L190)). But `opmath_type` only upcasts float16, bfloat16, and f8 to float. For float and double inputs, `acc_t` stays the same type ([OpMathType.h#L15-L27](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/OpMathType.h#L15-L27)). So while this fixes the intermediate precision problem for float16, the overflow problem in float itself is still there.

A standard fix for this is the one used in LAPACK's[ dnrm2](https://gensoft.pasteur.fr/docs/lapack/3.9.0/de/da4/group__double__blas__level1_ga4b25c539b862414e6f91ebb30b391d7c.html): scale by the max magnitude before summing.
```
def norm(a):
    t = torch.max(torch.abs(a))
    c = torch.abs(a / t) ** 2
    e = torch.sum(c)
    return torch.sqrt(e) * t
```

Dividing by the max keeps every term ≤ 1 before the sum, which avoids the overflow.

Does this fix look reasonable? @soulitzer  I’m happy to raise a PR.
